Create grain size evolution reaction model

gassmoeller opened the pull request:
This PR refactors the grain size material model and moves most of the code specific to grain size evolution into a reaction model called `GrainSizeEvolution`. This will make it much easier in the future to enhance the grain size material model with new functionality, or to use the grain size evolution functionality in other material models. This whole PR is essentially just a big move of code blocks from `source/material_model/grain_size.cc` to `source/material/reaction_model/grain_size_evolution.cc` and in the corresponding header files. There are minimal code changes to make this work, and all existing grain size tests except for one still show the same results.

The one exception is the test `grain_size_latent_heat`, which was badly designed from the start. It implements its own material model to achieve the desired analytical solution, but it does so by deriving from the grain size material model and then overriding the `evaluate` function. Besides the problem that this required to make a whole bunch of member variables in the `GrainSize` material model protected, it also meant the test used a different implementation than the material model itself, and so they have already diverged in their behavior. I removed part of these problems, now the test owns a member variable of type `GrainSize` material model, calls the `evaluate` function of the member variable and then overwrites the necessary material outputs afterwards. The reason the test output is slightly different now is that the test and the material model itself computed the density in slightly different ways. The test used the full pressure to compute the density, the material model uses the adiabatic pressure. It wasnt easy to keep the test behavior, and anyway we want the test to use as much functionality of the actual material model as possible, so I updated the test results. I made sure that the relevant output of the test (the temperature jump across a phase transition) is still correct.

The changes to the structure of the code and the changes to the test are separated into two separate commits to make review easier.

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+      {
+        double moment_of_grain_size_distribution (const unsigned int n)
+        {
+          // This function normalizes the grain size distribution using the nth moment.
```
```suggestion
          // This function calculates the nth moment of the lognormal distribution with mean = 0 and variance = 0.8.
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+        {
+          // This function normalizes the grain size distribution using the nth moment.
+          // Description can be found in eq 8 of Bercovici and Richard (2012)
+          // This is the variance of the log-normal distribution
```
```suggestion
          // The variance is hardcoded as 0.8 as described in that paper.
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+        double moment_of_grain_size_distribution (const unsigned int n)
+        {
+          // This function normalizes the grain size distribution using the nth moment.
+          // Description can be found in eq 8 of Bercovici and Richard (2012)
```
```suggestion
          // This function is used in the grain pinning model of Bercovici and Ricard (2012; Appendix F6).
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+                                typename Interface<dim>::MaterialModelOutputs       &out) const
+      {
+        // We want to iterate over the grain size evolution here, as we solve in fact an ordinary differential equation
+        // and it is not correct to use the starting grain size (and it introduces instabilities).
```
```suggestion
        // and it is not correct to use the starting grain size (which also introduces instabilities).
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+        data.initial_time = 0.0;
+        data.final_time = this->get_timestep();
+
+        // TODO: make this an input parameter.
```
Might be worth opening an issue to fix this as a starter project and cross-reference the issue number here.

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+
+            // reduce grain size to recrystallized_grain_size when crossing phase transitions
+            // if the distance in radial direction a grain moved compared to the last time step
+            // is crossing a phase transition, reduce grain size
```
```suggestion
            // indicates that it is crossing a phase transition.
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+                              "and likely an effect of a too large sub-timestep, or unrealistic "
+                              "input parameters."));
+
+            // reduce grain size to recrystallized_grain_size when crossing phase transitions
```
```suggestion
            // reduce grain size to recrystallized_grain_size
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+                  crossed_transition = phase;
+              }
+
+            // TODO: recrystallize first, and then do grain size growth/reduction for grains that crossed the transition
```
Again, perhaps worth opening a new issue if this hasn't been done yet.

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+      void
+      GrainSizeEvolution<dim>::parse_parameters (ParameterHandler &prm)
+      {
+        AssertThrow (this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::grain_size).size() == 1,
```
```suggestion
        AssertThrow (this->introspection().get_number_of_fields_of_type(CompositionalFieldDescription::grain_size) == 1,
```

bobmyhill commented on source/material_model/reaction_model/grain_size_evolution.cc:
```diff
@@ -0,0 +1,669 @@
+        AssertThrow (this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::grain_size).size() == 1,
+                     ExcMessage("The 'grain size' material model only works if exactly one compositional "
+                                "field with type 'grain size' is present. It looks like there are " +
+                                std::to_string(this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::grain_size).size())
```
```suggestion
                                std::to_string(this->introspection().get_number_of_fields_of_type(CompositionalFieldDescription::grain_size))
```

bobmyhill reviewed (changes requested):
Great, thanks @gassmoeller ! 

A few comments on documentation and description, then this is good to go.

gassmoeller commented:
> A few comments

Just a few comments :rofl:, but seriously thanks for the comments, parts of this code are 10 years old and needed an update. I updated all the documentation and the test.

> Might be worth opening an issue to fix this as a starter project and cross-reference the issue number here.

Hmm, I am not so sure about this. Personally, I think the first todo should not really be there. I do not see how anyone would have a way to pick a good starting timestep fraction.  We dont even know the numerical timestep well before we run a first model. I personally would just remove the first TODO.

And the second TODO would be nice to have, but is more complicated than the comment suggested. Instead of after the timestep (as implemented) or before the timestep (as the TODO suggested) we should  do the grain size reset *during* the grain size evolution, i.e. the SUNDIALS explicit function would need to check at every step if we crossed the transition. This may of course mess with the stability and convergence of the ODE solver, but if we go for more accuracy this would be the correct option. Whether we do the reset before or after the timestep probably affects the accuracy in about the same way.


I think I addressed everything, but feel free to take another look, my corrections are contained in the last commit.

bobmyhill commented on tests/grain_size_latent_heat.cc:
```diff
@@ -127,93 +117,63 @@ namespace aspect
+                  // compositional fields and position are not used for this test in the base model
+                  const double h = base_model->enthalpy(in.temperature[i],in.pressure[i],compositional_fields,Point<dim>());
+                  const double dh = base_model->enthalpy(in.temperature[i],in.pressure[i] + delta_pressure,compositional_fields,Point<dim>());
+                  dHdp = (dh - h) / delta_press;
```
```suggestion
                  dHdp = (dh - h) / delta_pressure;
```

bobmyhill reviewed (approved):
Brill, thanks! Apart from the one change (and perhaps the removal of TODO1), I think this is good to go.

TODO1: I guess my question would be is 0.001*timestep always a decent guess, or might it be good to allow the user to tweak the timestep fraction? Maybe this doesn't need a todo though - if it's ever a problem it'll be flagged as an issue.

TODO2: Argh, that sounds like a can of worms. Perhaps best to avoid.

gassmoeller commented:
Thanks, I fixed the test. Let's keep the TODO's in for now.
